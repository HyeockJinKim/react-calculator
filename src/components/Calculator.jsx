import * as React from "react";
import styled from "styled-components";

import Panel from "./Panel";
import Display from "./Display";
import ButtonGroup from "./ButtonGroup";
import Button from "./Button";
import History from "./History";

const Container = styled.div`
  margin: 30px auto;
  text-align: center;
`;

const Box = styled.div`
  display: inline-block;
  width: 270px;
  height: 65px;
  padding: 10px;
  border: 2px solid #000;
  border-radius: 5px;
  text-align: right;
  box-shadow: 1px 1px 3px rgba(0, 0, 0, 0.5);
  margin-bottom: 10px;
  cursor: pointer;
  h3 {
    margin: 0px;
  }
`;

const evalFunc = function(string) {
  // eslint-disable-next-line no-new-func
  return new Function("return (" + string + ")")();
};

class Calculator extends React.Component {
  state = {
    displayValue: "",
    history: []
  };

  onClickButton = key => {
    let { displayValue = "", history = [] } = this.state;
    displayValue = "" + displayValue;
    const lastChar = displayValue.substr(displayValue.length - 1);
    const operatorKeys = ["÷", "×", "-", "+"];
    const proc = {
      AC: () => {
        this.setState({ displayValue: "", history });
      },
      BS: () => {
        if (displayValue.length > 0) {
          displayValue = displayValue.substr(0, displayValue.length - 1);
        }
        this.setState({ displayValue, history });
      },
      "√": () => {
        let result = Math.sqrt(evalFunc(displayValue.replace(/÷/gi, "/").replace(/×/gi, "*").replace(/√/gi, "Math.sqrt")));
        history.splice(0, 0, {displayValue: "√("+displayValue+")", result});
        this.setState({ displayValue: result, history: [...history] })
      },
      "÷": () => {
        if (lastChar !== "" && !operatorKeys.includes(lastChar)) {
          this.setState({ displayValue: displayValue + "÷", history });
        }
      },
      "×": () => {
        if (lastChar !== "" && !operatorKeys.includes(lastChar)) {
          this.setState({ displayValue: displayValue + "×", history });
        }
      },
      "-": () => {
        if (lastChar !== "" && !operatorKeys.includes(lastChar)) {
          this.setState({ displayValue: displayValue + "-", history });
        }
      },
      "+": () => {
        if (lastChar !== "" && !operatorKeys.includes(lastChar)) {
          this.setState({ displayValue: displayValue + "+", history });
        }
      },
      "=": () => {
        if (lastChar !== "" && operatorKeys.includes(lastChar)) {
          displayValue = displayValue.substr(0, displayValue.length - 1);
        } else if (lastChar !== "") {
          let result = evalFunc(displayValue.replace(/÷/gi, "/").replace(/×/gi, "*").replace(/√/gi, "Math.sqrt"));
          history.splice(0, 0, {displayValue: displayValue, result});
          displayValue = result;
        }
        this.setState({ displayValue, history: [...history] });
      },
      ".": () => {
        if (lastChar !== "" && !operatorKeys.includes(lastChar) && lastChar !== ".") {
          this.setState({ displayValue: displayValue + ".", history });
        }
      },
      "0": () => {
        if (Number(displayValue) !== 0) {
          displayValue += "0";
          this.setState({ displayValue, history });
        }
      }
    };

    if (proc[key]) {
      proc[key]();
    } else {
      // 여긴 숫자
      this.setState({ displayValue: displayValue + key });
    }
  };

  render() {
    return (
      <Container>
        <Panel>
          <Display displayValue={this.state.displayValue} />
          <ButtonGroup onClickButton={this.onClickButton}>
            <Button size={1} color="gray">
              AC
            </Button>
            <Button size={1} color="gray">
              BS
            </Button>
            <Button size={1} color="gray">
              √
            </Button>
            <Button size={1} color="gray">
              ÷
            </Button>

            <Button size={1}>7</Button>
            <Button size={1}>8</Button>
            <Button size={1}>9</Button>
            <Button size={1} color="gray">
              ×
            </Button>

            <Button size={1}>4</Button>
            <Button size={1}>5</Button>
            <Button size={1}>6</Button>
            <Button size={1} color="gray">
              -
            </Button>

            <Button size={1}>1</Button>
            <Button size={1}>2</Button>
            <Button size={1}>3</Button>
            <Button size={1} color="gray">
              +
            </Button>

            <Button size={2}>0</Button>
            <Button size={1}>.</Button>
            <Button size={1} color="gray">
              =
            </Button>
          </ButtonGroup>
        </Panel>
        <History>
          {this.state.history.map((x, i) => (
            <Box onClick={() => this.setState({ displayValue: this.state.history[i].displayValue, history: this.state.history })}>
              <h3>{x.displayValue}</h3>
              <h3>{"= " + x.result}</h3>
            </Box>
          ))}
        </History>
      </Container>
    );
  }
}

export default Calculator;
