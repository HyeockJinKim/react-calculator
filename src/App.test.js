import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import { render, fireEvent, getByTestId, getByText } from '@testing-library/react';

import Calculator from "./components/Calculator";

describe('계산기 테스트', () => {
  it('정해진 계산기 모양과 같게 표시되는 지 테스트 (√ 버튼을 추가했는 지)', () => {
    const utils = render(<Calculator/>);
    expect(utils.container).toMatchSnapshot();
  });

  it('40 - 31', () => {
    let container = document.createElement('div');
    document.body.appendChild(container);
    act(() => {
      ReactDOM.render(<Calculator />, container);
    });
    let btn_4 = getByText(container, '4');
    let btn_0 = getByText(container, '0');
    let btn_3 = getByText(container, '3');
    let btn_1 = getByText(container, '1');
    let minus = getByText(container, '-');
    expect(getByTestId(container, 'display').textContent).toBe('');
    act(() => {
      fireEvent.click(btn_4);
      fireEvent.click(btn_0);
      fireEvent.click(minus);
      fireEvent.click(btn_3);
      fireEvent.click(btn_1);
    })

    expect(getByTestId(container, 'display')).toBe('40-31');
    document.body.removeChild(container);
  })
})
