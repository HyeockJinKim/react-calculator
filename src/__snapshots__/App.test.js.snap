// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`계산기 테스트 정해진 계산기 모양과 같게 표시되는 지 테스트 (√ 버튼을 추가했는 지) 1`] = `
<div>
  <div
    class="sc-EHOje fKXSMh"
  >
    <div
      class="sc-bdVaJa djxKOg"
    >
      <div
        class="sc-bwzfXH dLCcAN"
      >
        <div
          data-testid="display"
        />
      </div>
      <div
        class="sc-htpNat NNsK"
        data-testid="group"
      >
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button
            data-color="gray"
          >
            AC
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button
            data-color="gray"
          >
            BS
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button
            data-color="gray"
          >
            √
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button
            data-color="gray"
          >
            ÷
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            7
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            8
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            9
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button
            data-color="gray"
          >
            ×
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            4
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            5
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            6
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button
            data-color="gray"
          >
            -
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            1
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            2
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            3
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button
            data-color="gray"
          >
            +
          </button>
        </div>
        <div
          class="sc-bxivhb eTNCAz"
          size="2"
        >
          <button>
            0
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button>
            .
          </button>
        </div>
        <div
          class="sc-bxivhb dRcqyj"
          size="1"
        >
          <button
            data-color="gray"
          >
            =
          </button>
        </div>
      </div>
    </div>
    <div
      class="sc-ifAKCX hGKQdm"
    />
  </div>
</div>
`;
